### ISSUE TITLE
Feature suggestion: running tasks from non main/task threads

### ISSUE THREAD

**Issue by Vuhdo** [dougbinks/enkiTS#39]
Hi Doug,

enkiTS does not allow running tasks or waiting for completion from threads other than main/task threads, as I understood. 

For example, I would like to be able to use the system from rendering thread, which itself is not a task thread, but a full-fledged thread typically running in parallel with the main one. Or from background loading thread which is mostly idle waiting for IO, but uses tasks to decompress/finalize assets. 
It could be cool if enkiTS was able to support that. What do you think? Thanks!

-- Aleksei

**Comment by dougbinks** [comment]
There is a workaround for this, as it is possible to use the AddPinnedTask function to add a pinned task from any thread. So you can create and add a pinned task which adds the tasks you need.

This is what I do for IO threads. Note that you can also use AddPinnedTask to run your rendering function every frame on a given thread, and then this would be a tasking thread so would be able to add tasks normally.

It still makes sense to have a way to add general purpose tasks from any thread, I just need to work out the best approach.

**Comment by dougbinks** [comment]
I've updated the comments to show you can use AddPinnedTask from other threads in the latest release. This releases also clears the way for me to work on this feature - the tricky part is that it requires a MPMC queue, and I'd like to minimize both the amount of code and the impact on the scheduler performance (MPMC queues can be slow due to contention).

**Comment by Vuhdo** [comment]
Thanks for the clarification.

As far as I understand, there's no big difference between the main thread and non tasking thread. Since the number of such threads is quite low naturally, I imagined it would be enough to reserve more per-thread data (pipes, etc.) and register TLS data for those threads inside the `AddTaskSetToPipe()` call on  the fly. Even manually ad-hoc registering a thread would be quite ok I think.

**Comment by dougbinks** [comment]
I don't think I'd want to have this happen automatically, since enkiTS makes the guarantee that all allocations happen in user code or at initialization time. However one approach might be to add a RegisterThread() function which does what you suggest. I'll have a look at what that might entail.

**Comment by Vuhdo** [comment]
That would be great, thanks!

**Comment by dougbinks** [comment]
I just pushed [a branch called dev_registerthread](https://github.com/dougbinks/enkiTS/tree/dev_registerthread) which is the first draft of the interface plus workings, though the functionality is as yet untested.

There will be some performance work to do on optimizing some of the changes, but it looks like this approach could work.

I'm also considering making all non registered threads assert in debug mode to help find bugs.

Be aware this probably won't yet work - I'm about to create an example/test to find issues and fix.

**Comment by dougbinks** [comment]
I've just pushed an example for this, and so far the new functionality appears to be working.

If you get the chance I'd appreciate feedback, and in the meantime I will integrate this into [Avoyd](https://avoyd.com/) to help test prior to merging to master.

**Comment by dougbinks** [comment]
I've completed this development with examples and C interface, and it has now been integrated into the dev branch (and my project Avoyd) and so far works well.

I plan to merge to master soon and release.

**Comment by Vuhdo** [comment]
Awesome!

My plan is to replace old task scheduler in our game and see what happens. 
The game is relatively big, multiplatform (Windows/PS4/Xbox) and MMO, so I guess it may take some time.

**Comment by dougbinks** [comment]
I've now merged to master and published a release.

I'll close this issue but please re-open if you have problems.

Let me know how things go with the game. Sounds like something I might tweet about with https://twitter.com/indiegametech
